internal/backend/user_updates.go

```diff
@@ -15,89 +15,100 @@ import (
 
 	case *imap.MailboxUpdated:
-		return user.applyMailboxUpdated(ctx, tx, update)
+		return user.applyMailboxUpdated(ctx, update)
 
 	case *imap.MailboxIDChanged:
-		return user.applyMailboxIDChanged(ctx, tx, update)
+		return user.applyMailboxIDChanged(ctx, update)
 
 	case *imap.MessagesCreated:
-		return user.applyMessagesCreated(ctx, tx, update)
+		return user.applyMessagesCreated(ctx, update)
 
 	case *imap.MessageLabelsUpdated:
-		return user.applyMessageLabelsUpdated(ctx, tx, update)
+		return user.applyMessageLabelsUpdated(ctx, update)
 
 	case *imap.MessageFlagsUpdated:
-		return user.applyMessageFlagsUpdated(ctx, tx, update)
+		return user.applyMessageFlagsUpdated(ctx, update)
 
 	case *imap.MessageIDChanged:
-		return user.applyMessageIDChanged(ctx, tx, update)
+		return user.applyMessageIDChanged(ctx, update)
 
 	case *imap.MessageDeleted:
-		return user.applyMessageDeleted(ctx, tx, update)
+		return user.applyMessageDeleted(ctx, update)
 
 	default:
 		panic("bad update")
 	}
 }
 
 // applyMailboxCreated applies a MailboxCreated update.
-func (user *user) applyMailboxCreated(ctx context.Context, tx *ent.Tx, update *imap.MailboxCreated) error {
-	if exists, err := DBMailboxExistsWithRemoteID(ctx, tx.Client(), update.Mailbox.ID); err != nil {
+func (user *user) applyMailboxCreated(ctx context.Context, update *imap.MailboxCreated) error {
+	if exists, err := DBReadResult(ctx, user.db, func(ctx context.Context, client *ent.Client) (bool, error) {
+		return DBMailboxExistsWithRemoteID(ctx, client, update.Mailbox.ID)
+	}); err != nil {
 		return err
 	} else if exists {
 		return nil
 	}
 
 	internalMailboxID := imap.InternalMailboxID(uuid.NewString())
 
-	if _, err := DBCreateMailbox(
-		ctx,
-		tx,
-		internalMailboxID,
-		update.Mailbox.ID,
-		strings.Join(update.Mailbox.Name, user.delimiter),
-		update.Mailbox.Flags,
-		update.Mailbox.PermanentFlags,
-		update.Mailbox.Attributes,
-	); err != nil {
-		return err
-	}
+	return user.db.Write(ctx, func(ctx context.Context, tx *ent.Tx) error {
+		if _, err := DBCreateMailbox(
+			ctx,
+			tx,
+			internalMailboxID,
+			update.Mailbox.ID,
+			strings.Join(update.Mailbox.Name, user.delimiter),
+			update.Mailbox.Flags,
+			update.Mailbox.PermanentFlags,
+			update.Mailbox.Attributes,
+		); err != nil {
+			return err
+		}
 
-	return nil
+		return nil
+	})
 }
 
 // applyMailboxDeleted applies a MailboxDeleted update.
-func (user *user) applyMailboxDeleted(ctx context.Context, tx *ent.Tx, update *imap.MailboxDeleted) error {
-	if exists, err := DBMailboxExistsWithRemoteID(ctx, tx.Client(), update.MailboxID); err != nil {
+func (user *user) applyMailboxDeleted(ctx context.Context, update *imap.MailboxDeleted) error {
+	if exists, err := DBReadResult(ctx, user.db, func(ctx context.Context, client *ent.Client) (bool, error) {
+		return DBMailboxExistsWithRemoteID(ctx, client, update.MailboxID)
+	}); err != nil {
 		return err
-	} else if !exists {
+	} else if exists {
```

Comment:
still in need of a decent morning coffee, but why did you change this from `!exists` to `exists`...? was this a bug you fixed, or did you make a mistake here?

Reply:
Good catch. Fixed.